\documentclass[11pt]{article}
\usepackage{sectsty}
\allsectionsfont{\color{blue}\fontfamily{lmss}\selectfont}
\usepackage{fontspec}
\setmainfont{XCharter}

\usepackage{listings}
\lstset{
basicstyle=\small\ttfamily,
tabsize=8,
columns=flexible,
breaklines=true,
frame=tb,
rulecolor=\color[rgb]{0.8,0.8,0.7},
backgroundcolor=\color[rgb]{1,1,0.91},
postbreak=\raisebox{0ex}[0ex][0ex]{\ensuremath{\color{red}\hookrightarrow\space}}
}
\usepackage{fontawesome}


\usepackage{mdframed}
\newmdenv[
  backgroundcolor=gray,
  fontcolor=white,
  nobreak=true,
]{terminalinput}



\usepackage{parskip}


    \usepackage[breakable]{tcolorbox}
    \usepackage{parskip} % Stop auto-indenting (to mimic markdown behaviour)

    \usepackage{iftex}
    \ifPDFTeX
    	\usepackage[T1]{fontenc}
    	\usepackage{mathpazo}
    \else
    	\usepackage{fontspec}
    \fi

    % Basic figure setup, for now with no caption control since it's done
    % automatically by Pandoc (which extracts ![](path) syntax from Markdown).
    \usepackage{graphicx}
    % Maintain compatibility with old templates. Remove in nbconvert 6.0
    \let\Oldincludegraphics\includegraphics
    % Ensure that by default, figures have no caption (until we provide a
    % proper Figure object with a Caption API and a way to capture that
    % in the conversion process - todo).
    \usepackage{caption}
    \DeclareCaptionFormat{nocaption}{}
    \captionsetup{format=nocaption,aboveskip=0pt,belowskip=0pt}

    \usepackage[Export]{adjustbox} % Used to constrain images to a maximum size
    \adjustboxset{max size={0.9\linewidth}{0.9\paperheight}}
    \usepackage{float}
    \floatplacement{figure}{H} % forces figures to be placed at the correct location
    \usepackage{xcolor} % Allow colors to be defined
    \usepackage{enumerate} % Needed for markdown enumerations to work
    \usepackage{geometry} % Used to adjust the document margins
    \usepackage{amsmath} % Equations
    \usepackage{amssymb} % Equations
    \usepackage{textcomp} % defines textquotesingle
    % Hack from http://tex.stackexchange.com/a/47451/13684:
    \AtBeginDocument{%
        \def\PYZsq{\textquotesingle}% Upright quotes in Pygmentized code
    }
    \usepackage{upquote} % Upright quotes for verbatim code
    \usepackage{eurosym} % defines \euro
    \usepackage[mathletters]{ucs} % Extended unicode (utf-8) support
    \usepackage{fancyvrb} % verbatim replacement that allows latex

    % The hyperref package gives us a pdf with properly built
    % internal navigation ('pdf bookmarks' for the table of contents,
    % internal cross-reference links, web links for URLs, etc.)
    \usepackage{hyperref}
    % The default LaTeX title has an obnoxious amount of whitespace. By default,
    % titling removes some of it. It also provides customization options.
    \usepackage{titling}
    \usepackage{longtable} % longtable support required by pandoc >1.10
    \usepackage{booktabs}  % table support for pandoc > 1.12.2
    \usepackage[inline]{enumitem} % IRkernel/repr support (it uses the enumerate* environment)
    \usepackage[normalem]{ulem} % ulem is needed to support strikethroughs (\sout)
                                % normalem makes italics be italics, not underlines
    \usepackage{mathrsfs}



    % Colors for the hyperref package
    \definecolor{urlcolor}{rgb}{0,.145,.698}
    \definecolor{linkcolor}{rgb}{.71,0.21,0.01}
    \definecolor{citecolor}{rgb}{.12,.54,.11}

    % ANSI colors
    \definecolor{ansi-black}{HTML}{3E424D}
    \definecolor{ansi-black-intense}{HTML}{282C36}
    \definecolor{ansi-red}{HTML}{E75C58}
    \definecolor{ansi-red-intense}{HTML}{B22B31}
    \definecolor{ansi-green}{HTML}{00A250}
    \definecolor{ansi-green-intense}{HTML}{007427}
    \definecolor{ansi-yellow}{HTML}{DDB62B}
    \definecolor{ansi-yellow-intense}{HTML}{B27D12}
    \definecolor{ansi-blue}{HTML}{208FFB}
    \definecolor{ansi-blue-intense}{HTML}{0065CA}
    \definecolor{ansi-magenta}{HTML}{D160C4}
    \definecolor{ansi-magenta-intense}{HTML}{A03196}
    \definecolor{ansi-cyan}{HTML}{60C6C8}
    \definecolor{ansi-cyan-intense}{HTML}{258F8F}
    \definecolor{ansi-white}{HTML}{C5C1B4}
    \definecolor{ansi-white-intense}{HTML}{A1A6B2}
    \definecolor{ansi-default-inverse-fg}{HTML}{FFFFFF}
    \definecolor{ansi-default-inverse-bg}{HTML}{000000}

    % commands and environments needed by pandoc snippets
    % extracted from the output of `pandoc -s`
    \providecommand{\tightlist}{%
      \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
    \DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
    % Add ',fontsize=\small' for more characters per line
    \newenvironment{Shaded}{}{}
    \newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.56,0.13,0.00}{{#1}}}
    \newcommand{\DecValTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\FloatTok}[1]{\textcolor[rgb]{0.25,0.63,0.44}{{#1}}}
    \newcommand{\CharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\StringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\CommentTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textit{{#1}}}}
    \newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{{#1}}}
    \newcommand{\AlertTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.02,0.16,0.49}{{#1}}}
    \newcommand{\RegionMarkerTok}[1]{{#1}}
    \newcommand{\ErrorTok}[1]{\textcolor[rgb]{1.00,0.00,0.00}{\textbf{{#1}}}}
    \newcommand{\NormalTok}[1]{{#1}}

    % Additional commands for more recent versions of Pandoc
    \newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.53,0.00,0.00}{{#1}}}
    \newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.25,0.44,0.63}{{#1}}}
    \newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.73,0.40,0.53}{{#1}}}
    \newcommand{\ImportTok}[1]{{#1}}
    \newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.73,0.13,0.13}{\textit{{#1}}}}
    \newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\VariableTok}[1]{\textcolor[rgb]{0.10,0.09,0.49}{{#1}}}
    \newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.44,0.13}{\textbf{{#1}}}}
    \newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.40,0.40,0.40}{{#1}}}
    \newcommand{\BuiltInTok}[1]{{#1}}
    \newcommand{\ExtensionTok}[1]{{#1}}
    \newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.74,0.48,0.00}{{#1}}}
    \newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.49,0.56,0.16}{{#1}}}
    \newcommand{\InformationTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}
    \newcommand{\WarningTok}[1]{\textcolor[rgb]{0.38,0.63,0.69}{\textbf{\textit{{#1}}}}}


    % Define a nice break command that doesn't care if a line doesn't already
    % exist.
    \def\br{\hspace*{\fill} \\* }
    % Math Jax compatibility definitions
    \def\gt{>}
    \def\lt{<}
    \let\Oldtex\TeX
    \let\Oldlatex\LaTeX
    \renewcommand{\TeX}{\textrm{\Oldtex}}
    \renewcommand{\LaTeX}{\textrm{\Oldlatex}}
    % Document parameters
    % Document title
    \title{index}





% Pygments definitions
\makeatletter
\def\PY@reset{\let\PY@it=\relax \let\PY@bf=\relax%
    \let\PY@ul=\relax \let\PY@tc=\relax%
    \let\PY@bc=\relax \let\PY@ff=\relax}
\def\PY@tok#1{\csname PY@tok@#1\endcsname}
\def\PY@toks#1+{\ifx\relax#1\empty\else%
    \PY@tok{#1}\expandafter\PY@toks\fi}
\def\PY@do#1{\PY@bc{\PY@tc{\PY@ul{%
    \PY@it{\PY@bf{\PY@ff{#1}}}}}}}
\def\PY#1#2{\PY@reset\PY@toks#1+\relax+\PY@do{#2}}

\expandafter\def\csname PY@tok@w\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.73,0.73}{##1}}}
\expandafter\def\csname PY@tok@c\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.74,0.48,0.00}{##1}}}
\expandafter\def\csname PY@tok@k\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.69,0.00,0.25}{##1}}}
\expandafter\def\csname PY@tok@o\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ow\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@nb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@nn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@ne\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.82,0.25,0.23}{##1}}}
\expandafter\def\csname PY@tok@nv\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@no\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@nl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@ni\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.60,0.60,0.60}{##1}}}
\expandafter\def\csname PY@tok@na\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.49,0.56,0.16}{##1}}}
\expandafter\def\csname PY@tok@nt\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@nd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.67,0.13,1.00}{##1}}}
\expandafter\def\csname PY@tok@s\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sd\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@si\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@se\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.13}{##1}}}
\expandafter\def\csname PY@tok@sr\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.40,0.53}{##1}}}
\expandafter\def\csname PY@tok@ss\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sx\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@m\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@gh\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gu\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.50,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@gd\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.63,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@gi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.63,0.00}{##1}}}
\expandafter\def\csname PY@tok@gr\endcsname{\def\PY@tc##1{\textcolor[rgb]{1.00,0.00,0.00}{##1}}}
\expandafter\def\csname PY@tok@ge\endcsname{\let\PY@it=\textit}
\expandafter\def\csname PY@tok@gs\endcsname{\let\PY@bf=\textbf}
\expandafter\def\csname PY@tok@gp\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,0.50}{##1}}}
\expandafter\def\csname PY@tok@go\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.53,0.53,0.53}{##1}}}
\expandafter\def\csname PY@tok@gt\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.27,0.87}{##1}}}
\expandafter\def\csname PY@tok@err\endcsname{\def\PY@bc##1{\setlength{\fboxsep}{0pt}\fcolorbox[rgb]{1.00,0.00,0.00}{1,1,1}{\strut ##1}}}
\expandafter\def\csname PY@tok@kc\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kd\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kn\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@kr\endcsname{\let\PY@bf=\textbf\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@bp\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.50,0.00}{##1}}}
\expandafter\def\csname PY@tok@fm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.00,0.00,1.00}{##1}}}
\expandafter\def\csname PY@tok@vc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vg\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@vm\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.10,0.09,0.49}{##1}}}
\expandafter\def\csname PY@tok@sa\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sc\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@dl\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s2\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@sh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@s1\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.73,0.13,0.13}{##1}}}
\expandafter\def\csname PY@tok@mb\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mf\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mh\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mi\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@il\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@mo\endcsname{\def\PY@tc##1{\textcolor[rgb]{0.40,0.40,0.40}{##1}}}
\expandafter\def\csname PY@tok@ch\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cm\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cpf\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@c1\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}
\expandafter\def\csname PY@tok@cs\endcsname{\let\PY@it=\textit\def\PY@tc##1{\textcolor[rgb]{0.25,0.50,0.50}{##1}}}

\def\PYZbs{\char`\\}
\def\PYZus{\char`\_}
\def\PYZob{\char`\{}
\def\PYZcb{\char`\}}
\def\PYZca{\char`\^}
\def\PYZam{\char`\&}
\def\PYZlt{\char`\<}
\def\PYZgt{\char`\>}
\def\PYZsh{\char`\#}
\def\PYZpc{\char`\%}
\def\PYZdl{\char`\$}
\def\PYZhy{\char`\-}
\def\PYZsq{\char`\'}
\def\PYZdq{\char`\"}
\def\PYZti{\char`\~}
% for compatibility with earlier versions
\def\PYZat{@}
\def\PYZlb{[}
\def\PYZrb{]}
\makeatother


    % For linebreaks inside Verbatim environment from package fancyvrb.
    \makeatletter
        \newbox\Wrappedcontinuationbox
        \newbox\Wrappedvisiblespacebox
        \newcommand*\Wrappedvisiblespace {\textcolor{red}{\textvisiblespace}}
        \newcommand*\Wrappedcontinuationsymbol {\textcolor{red}{\llap{\tiny$\m@th\hookrightarrow$}}}
        \newcommand*\Wrappedcontinuationindent {3ex }
        \newcommand*\Wrappedafterbreak {\kern\Wrappedcontinuationindent\copy\Wrappedcontinuationbox}
        % Take advantage of the already applied Pygments mark-up to insert
        % potential linebreaks for TeX processing.
        %        {, <, #, %, $, ' and ": go to next line.
        %        _, }, ^, &, >, - and ~: stay at end of broken line.
        % Use of \textquotesingle for straight quote.
        \newcommand*\Wrappedbreaksatspecials {%
            \def\PYGZus{\discretionary{\char`\_}{\Wrappedafterbreak}{\char`\_}}%
            \def\PYGZob{\discretionary{}{\Wrappedafterbreak\char`\{}{\char`\{}}%
            \def\PYGZcb{\discretionary{\char`\}}{\Wrappedafterbreak}{\char`\}}}%
            \def\PYGZca{\discretionary{\char`\^}{\Wrappedafterbreak}{\char`\^}}%
            \def\PYGZam{\discretionary{\char`\&}{\Wrappedafterbreak}{\char`\&}}%
            \def\PYGZlt{\discretionary{}{\Wrappedafterbreak\char`\<}{\char`\<}}%
            \def\PYGZgt{\discretionary{\char`\>}{\Wrappedafterbreak}{\char`\>}}%
            \def\PYGZsh{\discretionary{}{\Wrappedafterbreak\char`\#}{\char`\#}}%
            \def\PYGZpc{\discretionary{}{\Wrappedafterbreak\char`\%}{\char`\%}}%
            \def\PYGZdl{\discretionary{}{\Wrappedafterbreak\char`\$}{\char`\$}}%
            \def\PYGZhy{\discretionary{\char`\-}{\Wrappedafterbreak}{\char`\-}}%
            \def\PYGZsq{\discretionary{}{\Wrappedafterbreak\textquotesingle}{\textquotesingle}}%
            \def\PYGZdq{\discretionary{}{\Wrappedafterbreak\char`\"}{\char`\"}}%
            \def\PYGZti{\discretionary{\char`\~}{\Wrappedafterbreak}{\char`\~}}%
        }
        % Some characters . , ; ? ! / are not pygmentized.
        % This macro makes them "active" and they will insert potential linebreaks
        \newcommand*\Wrappedbreaksatpunct {%
            \lccode`\~`\.\lowercase{\def~}{\discretionary{\hbox{\char`\.}}{\Wrappedafterbreak}{\hbox{\char`\.}}}%
            \lccode`\~`\,\lowercase{\def~}{\discretionary{\hbox{\char`\,}}{\Wrappedafterbreak}{\hbox{\char`\,}}}%
            \lccode`\~`\;\lowercase{\def~}{\discretionary{\hbox{\char`\;}}{\Wrappedafterbreak}{\hbox{\char`\;}}}%
            \lccode`\~`\:\lowercase{\def~}{\discretionary{\hbox{\char`\:}}{\Wrappedafterbreak}{\hbox{\char`\:}}}%
            \lccode`\~`\?\lowercase{\def~}{\discretionary{\hbox{\char`\?}}{\Wrappedafterbreak}{\hbox{\char`\?}}}%
            \lccode`\~`\!\lowercase{\def~}{\discretionary{\hbox{\char`\!}}{\Wrappedafterbreak}{\hbox{\char`\!}}}%
            \lccode`\~`\/\lowercase{\def~}{\discretionary{\hbox{\char`\/}}{\Wrappedafterbreak}{\hbox{\char`\/}}}%
            \catcode`\.\active
            \catcode`\,\active
            \catcode`\;\active
            \catcode`\:\active
            \catcode`\?\active
            \catcode`\!\active
            \catcode`\/\active
            \lccode`\~`\~
        }
    \makeatother

    \let\OriginalVerbatim=\Verbatim
    \makeatletter
    \renewcommand{\Verbatim}[1][1]{%
        %\parskip\z@skip
        \sbox\Wrappedcontinuationbox {\Wrappedcontinuationsymbol}%
        \sbox\Wrappedvisiblespacebox {\FV@SetupFont\Wrappedvisiblespace}%
        \def\FancyVerbFormatLine ##1{\hsize\linewidth
            \vtop{\raggedright\hyphenpenalty\z@\exhyphenpenalty\z@
                \doublehyphendemerits\z@\finalhyphendemerits\z@
                \strut ##1\strut}%
        }%
        % If the linebreak is at a space, the latter will be displayed as visible
        % space at end of first line, and a continuation symbol starts next line.
        % Stretch/shrink are however usually zero for typewriter font.
        \def\FV@Space {%
            \nobreak\hskip\z@ plus\fontdimen3\font minus\fontdimen4\font
            \discretionary{\copy\Wrappedvisiblespacebox}{\Wrappedafterbreak}
            {\kern\fontdimen2\font}%
        }%

        % Allow breaks at special characters using \PYG... macros.
        \Wrappedbreaksatspecials
        % Breaks at punctuation characters . , ; ? ! and / need catcode=\active
        \OriginalVerbatim[#1,codes*=\Wrappedbreaksatpunct]%
    }
    \makeatother

    % Exact colors from NB
    \definecolor{incolor}{HTML}{303F9F}
    \definecolor{outcolor}{HTML}{D84315}
    \definecolor{cellborder}{HTML}{CFCFCF}
    \definecolor{cellbackground}{HTML}{F7F7F7}

    % prompt
    \makeatletter
    \newcommand{\boxspacing}{\kern\kvtcb@left@rule\kern\kvtcb@boxsep}
    \makeatother
    \newcommand{\prompt}[4]{
        {\ttfamily\llap{{\color{blue}\LARGE\faKeyboardO\hspace{3pt}#4}}\vspace{-\baselineskip}}
    }



    % Prevent overflowing lines due to hard-to-break entities
    \sloppy
    % Setup hyperref package
    \hypersetup{
      breaklinks=true,  % so long urls are correctly broken across lines
      colorlinks=true,
      urlcolor=urlcolor,
      linkcolor=linkcolor,
      citecolor=citecolor,
      }
    % Slightly bigger margins than the latex defaults

    \geometry{verbose,tmargin=1in,bmargin=1in,lmargin=1in,rmargin=1in}



\renewcommand{\PY}[2]{{#2}}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{\color{gray}\sf\small\rightmark}
\lhead{\nouppercase{\color{gray}\sf\small\leftmark}}
\cfoot{\color{gray}\sf\thepage}
\renewcommand{\footrulewidth}{1pt}
\begin{document}





    \hypertarget{antimicrobial-resistance-identification-using-ariba}{%
\section{Antimicrobial resistance identification using
ARIBA}\label{antimicrobial-resistance-identification-using-ariba}}

\hypertarget{introduction}{%
\subsection{Introduction}\label{introduction}}

ARIBA is a tool that can be used to identify the prescence/abscence of a
set of genes and their variants in a sample. In particular, it was
designed to detect and identify antibiotic resistance genes. This
tutorial will walk you through the analysis of the \textit{Neisseria
gonorrhoeae} data set used in the ARIBA paper:

\begin{quote}
\textbf{ARIBA: rapid antimicrobial resistance genotyping directly from
sequencing reads}\\
Hunt M, Mather AE, Sánchez-Busó L, Page AJ, Parkhill J , Keane JA,
Harris SR.\\
\textit{Microbial Genomics 2017. doi: 110.1099/mgen.0.000131}\\
PMID:
\href{https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5695208/}{29177089}
\end{quote}

A copy of the paper can be found at

\texttt{\textasciitilde{}/course\_data/amr/ariba\_paper.pdf}

\hypertarget{learning-outcomes}{%
\subsection{Learning outcomes}\label{learning-outcomes}}

By the end of this tutorial you can expect to be able to:

\begin{itemize}
\tightlist
\item
  Use ARIBA to detect the prescence/abscence of a set of genes
\item
  Download and prepare the standard AMR databases for use with ARIBA
\item
  Run ARIBA on several samples to identify antibiotic resistance
\item
  Prepare and use your own database for use with ARIBA
\item
  Summarise ARIBA results for several samples
\item
  Query the AMR results produced by ARIBA
\item
  Use Phandango to visualise ARIBA results
\end{itemize}

\hypertarget{tutorial-sections}{%
\subsection{Tutorial sections}\label{tutorial-sections}}

This tutorial comprises the following sections:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  \href{basic_ariba.ipynb}{Detect prescence/abscence of a set of genes
  with ARIBA}
\item
  \href{standard_ariba.ipynb}{Use a standard AMR database with ARIBA}
\item
  \href{make_custom_db.ipynb}{Prepare a custom reference database for
  ARIBA}
\item
  \href{run_ariba.ipynb}{Run ARIBA using a custom reference database}
\item
  \href{phandango.ipynb}{View summarized results using Phandango}
\item
  \href{micplot.ipynb}{Investigate MIC data in relation to variants in
  the samples}
\end{enumerate}

\hypertarget{authors-and-license}{%
\subsection{Authors and License}\label{authors-and-license}}

This tutorial was created by
\href{https://github.com/jacquikeane}{Jacqui Keane} and
\href{https://github.com/martinghunt}{Martin Hunt}.

The content is licensed under a
\href{https://creativecommons.org/licenses/by/4.0/}{Creative Commons
Attribution 4.0 International License (CC-By 4.0)}.

    \hypertarget{running-commands-in-this-tutorial}{%
\subsubsection{Running commands in this
tutorial}\label{running-commands-in-this-tutorial}}

You can follow this tutorial by typing all the commands you see in a
terminal window on your computer. This is similar to the ``Command
Prompt'' window on MS Windows systems, which allows the user to type DOS
commands to manage files.

To get started, open a terminal window and type the command below
followed by the \texttt{Enter} key:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} \PYZti{}/course\PYZus{}data/amr/data
\end{Verbatim}
\end{tcolorbox}

    Now you can follow the instructions in the tutorial from here.

\hypertarget{prerequisites}{%
\subsection{Prerequisites}\label{prerequisites}}

This tutorial assumes that you have the \texttt{ARIBA} software
application installed on your computer. It has already been installed on
the computer you are using for this training course. To activate the
ARIBA conda environment and check that the software is installed
correctly, run the following command:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
conda activate ariba\PYZhy{}2.14.6
ariba \PYZhy{}\PYZhy{}help
\end{Verbatim}
\end{tcolorbox}

    This should return the following help message:

\begin{verbatim}
usage: ariba <command> <options>

ARIBA: Antibiotic Resistance Identification By Assembly

optional arguments:
  -h, --help      show this help message and exit

Available commands:

    aln2meta      Converts multi-aln fasta and SNPs to metadata
    expandflag    Expands flag column of report file
    flag          Translate the meaning of a flag
    getref        Download reference data
    micplot       Make violin/dot plots using MIC data
    prepareref    Prepare reference data for input to "run"
    pubmlstget    Download species from PubMLST and make db
    pubmlstspecies
                  Get list of available species from PubMLST
    refquery      Get cluster or sequence info from prepareref output
    run           Run the local assembly pipeline
    summary       Summarise multiple reports made by "run"
    test          Run small built-in test dataset
    version       Get versions and exit
\end{verbatim}

For download and installation instructions for ARIBA, please see the
\href{https://github.com/sanger-pathogens/ariba}{ARIBA GitHub-page at
https://github.com/sanger-pathogens/ariba}.

To get started with the tutorial, head to the first section:
\href{basic_ariba.ipynb}{Detect prescence/abscence of genes with ARIBA}


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{detect-prescenceabscence-of-genes-with-ariba}{%
\section{Detect prescence/abscence of genes with
ARIBA}\label{detect-prescenceabscence-of-genes-with-ariba}}

One of the most basic tasks ARIBA can be used for is to detect the
presence or abscence of a set genes. Let us look at an example. But
first, let's check that you're in the right place. Type the command
below in the terminal window followed by the Enter key:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{pwd}
\end{Verbatim}
\end{tcolorbox}

    It should display something like:

\texttt{/home/manager/course\_data/amr/data}

    Ok, back to the example! We have sequence data for 3 \textbf{Neisseria
gonorrhoeae} samples and want to determine if the genes \texttt{fitA}
and \texttt{fbpA} are present or absent in these samples. Let us use
ARIBA to determine this from the raw sequence data.

First list the fastq files for the samples.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls fastq
\end{Verbatim}
\end{tcolorbox}

    For the purpose of this exercise, we have created a fasta file
containing the sequences of the genes \texttt{fitA} and \texttt{fbpA},
take a look:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} basic
ls
cat genes.fasta
\end{Verbatim}
\end{tcolorbox}

    Running ARIBA is generally a three stage process: * prepare a reference
database * run ariba on each sample * summmarise the ariba results for
all samples

    First prepare the reference database:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba prepareref \PYZhy{}f genes.fasta \PYZhy{}\PYZhy{}all\PYZus{}coding yes out
\end{Verbatim}
\end{tcolorbox}

    Next, run ariba for all 3 samples

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba run out ../fastq/ERR1067813\PYZus{}1.fastq.gz ../fastq/ERR1067813\PYZus{}2.fastq.gz \PY{l+s+se}{\PYZbs{}}
ariba\PYZus{}results\PYZus{}ERR1067813
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba run out ../fastq/ERR1067814\PYZus{}1.fastq.gz ../fastq/ERR1067814\PYZus{}2.fastq.gz \PY{l+s+se}{\PYZbs{}}
ariba\PYZus{}results\PYZus{}ERR1067814
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba run out ../fastq/ERR1067815\PYZus{}1.fastq.gz ../fastq/ERR1067815\PYZus{}2.fastq.gz \PY{l+s+se}{\PYZbs{}}
ariba\PYZus{}results\PYZus{}ERR1067815
\end{Verbatim}
\end{tcolorbox}

    Next, summarise the results (the \texttt{-\/-no\_tree} option tells
ARIBA not to generate a tree):

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}no\PYZus{}tree all\PYZus{}samples ariba\PYZus{}results*/report.tsv
\end{Verbatim}
\end{tcolorbox}

    Now inspect the results:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat all\PYZus{}samples.csv
\end{Verbatim}
\end{tcolorbox}

    Notice that the original gene names are not present in the summary file,
instead we have the cluster names cluster, cluster\_1 etc. This is
because ARIBA gathers together genes that are similar into clusters.

To determine which genes are in which cluster run:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba refquery out cluster cluster
ariba refquery out cluster cluster\PYZus{}1
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  How many samples have the fitA gene?
\item
  How many samples have the fbpA gene?
\end{enumerate}

    Now go to the next part of the tutorial where we
\href{standard_ariba.ipynb}{use a standard database with ARIBA to
determine AMR}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{use-a-standard-amr-database-with-ariba}{%
\section{Use a standard AMR database with
ARIBA}\label{use-a-standard-amr-database-with-ariba}}

This sections shows you how to run ARIBA on several samples to determine
the presence of AMR genes using standard public AMR databases.

    Make a new directory \texttt{standard} and move into that directory:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ..
mkdir standard
\PY{n+nb}{cd} standard
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{reference-database}{%
\subsection{Reference database}\label{reference-database}}

We are going to download and prepare a reference database for use with
ARIBA. You can use any one of the
\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-getref}{standard
reference databases that ARIBA supports}

\begin{itemize}
\tightlist
\item
  ARG-ANNOT. PMID: 24145532
\item
  CARD. PMID: 23650175
\item
  MEGARes PMID: 27899569
\item
  NCBI BioProject: PRJNA313047
\item
  plasmidfinder PMID: 24777092
\item
  resfinder. PMID: 22782487
\item
  VFDB. PMID: 26578559
\item
  SRST2's version of ARG-ANNOT. PMID: 25422674.
\item
  VirulenceFinder PMID: 24574290.
\end{itemize}

Let's use the CARD database. The CARD (Comprehensive Antibiotic
Resistance Database) is a biological database that collects and
organizes reference information on antimicrobial resistance genes,
proteins and phenotypes. Type the following commands to make an ARIBA
database directory called \texttt{ariba\_card\_db}:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba getref card out.card
ariba prepareref \PYZhy{}f out.card.fa \PYZhy{}m out.card.tsv ariba\PYZus{}card\PYZus{}db
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{run-ariba-on-one-sample}{%
\subsection{Run ARIBA on one sample}\label{run-ariba-on-one-sample}}

To run ARIBA on one of the \textit{Neisseria gonorrhoeae} samples using
the CARD database, you need to supply the database directory we prepared
earlier (which we have called \texttt{ariba\_card\_db}) and two
sequencing reads files \texttt{fastq/ERR1067813\_1.fastq.gz},
\texttt{fastq/ERR1067813\_2.fastq.gz} along with the name of the output
directory \texttt{ERR1067813}.

The command to run ARIBA is:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba run ariba\PYZus{}card\PYZus{}db ../fastq/ERR1067813\PYZus{}1.fastq.gz \PY{l+s+se}{\PYZbs{}}
../fastq/ERR1067813\PYZus{}2.fastq.gz ERR1067813
\end{Verbatim}
\end{tcolorbox}

    This may take a few minutes to run. When complete, check what files and
directories were created by ARIBA:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    You should see a new directory called \texttt{ERR1067813} that will
contain the ARIBA results.

    \hypertarget{run-ariba-on-all-samples}{%
\subsection{Run ARIBA on all samples}\label{run-ariba-on-all-samples}}

But what if we want to run ARIBA on all 3 samples? This can be done with
a ``for'' loop. We assume that the reads files are named like this:

\begin{verbatim}
ERR1067813_1.fastq.gz ERR1067813_2.fastq.gz
ERR1067814_1.fastq.gz ERR1067814_2.fastq.gz
ERR1067815_1.fastq.gz ERR1067815_2.fastq.gz
\end{verbatim}

To run ARIBA on all samples type:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{k}{for} s in \PY{l+s+sb}{`}ls ../fastq/*1.f*.gz \PY{p}{|} sed \PY{l+s+s1}{\PYZsq{}s/..\PYZbs{}/fastq\PYZbs{}///\PYZsq{}} \PY{p}{|} sed \PY{l+s+s1}{\PYZsq{}s/\PYZbs{}\PYZus{}1.fastq.gz//\PYZsq{}}\PY{l+s+sb}{`}
\PY{k}{do}
    ariba run ariba\PYZus{}card\PYZus{}db ../fastq/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{s}\PY{l+s+si}{\PYZcb{}}\PYZus{}1.fastq.gz ../fastq/\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{s}\PY{l+s+si}{\PYZcb{}}\PYZus{}2.fastq.gz \PY{l+s+se}{\PYZbs{}}
    ariba.\PY{l+s+si}{\PYZdl{}\PYZob{}}\PY{n+nv}{s}\PY{l+s+si}{\PYZcb{}}
\PY{k}{done}
\end{Verbatim}
\end{tcolorbox}

    The above command may take a few mintes to run. While you are waiting
for it to finish read the explanation of the for loop below.

The loop iterates through each item returned by the command
\texttt{ls\ ../fastq/*1.f*.gz\ \textbar{}\ sed\ \textquotesingle{}s/..\textbackslash{}/fastq\textbackslash{}///\textquotesingle{}\ \textbar{}\ sed\ \textquotesingle{}s/\textbackslash{}\_1.fastq.gz//\textquotesingle{}}
and assigns the value of the item to the variable \texttt{\$s}.

The body of the loop runs the command \texttt{ariba\ run}. The options
passed to \texttt{ariba\ run} are the fastq files for the sample
specified as \texttt{../fastq/\$\{s\}\_1.fastq.gz} and
\texttt{../fastq/\$\{s\}\_2.fastq.gz} and the output directory for the
sample specified as \texttt{ariba.\$s}. For example, for sample
ERR1067813, \$s is ERR1067813 and the fastq files are
../fastq/ERR1067813\_1.fastq.gz and ../fastq/ERR1067813\_2.fastq.gz and
ariba.ERR1067813 is the output directory.

Let's also look more closely at the command
\texttt{ls\ ../fastq/*1.f*.gz\ \textbar{}\ sed\ \textquotesingle{}s/..\textbackslash{}/fastq\textbackslash{}///\textquotesingle{}\ \textbar{}\ sed\ \textquotesingle{}s/\textbackslash{}\_1.fastq.gz//\textquotesingle{}}.

    First try:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ../fastq/*1.f*.gz
\end{Verbatim}
\end{tcolorbox}

    This lists all the files in the \texttt{../fastq} directory that end
with `1.fastq.gz' which should give us one fastq file for each sample.

However all the files are prefixed with the path or location of the file
(\texttt{../fastq/}). To remove this we will use the \texttt{sed}
command to replace all occurences of \texttt{../fastq/} with an empty
string ''. Note the additional backslashes (\textbackslash{}) used to
escape the forward slashes (/) in the path.

Try it:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ../fastq/*1.f*.gz \PY{p}{|} sed \PY{l+s+s1}{\PYZsq{}s/..\PYZbs{}/fastq\PYZbs{}///\PYZsq{}}
\end{Verbatim}
\end{tcolorbox}

    The above command now lists just the fastq files which include the
\_1.fastq.gz at the end. Our aim is to extract a sample name or unique
identifier for each of our samples (so everything before the
\_1.fastq.gz). To acieve this we use an additional \texttt{sed} to
replace the occurence of \texttt{\_1.fastq.gz} with an empty string ``.

Try it:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ../fastq/*1.fastq.gz \PY{p}{|} sed \PY{l+s+s1}{\PYZsq{}s/..\PYZbs{}/fastq\PYZbs{}///\PYZsq{}} \PY{p}{|} sed \PY{l+s+s1}{\PYZsq{}s/\PYZbs{}\PYZus{}1.fastq.gz//\PYZsq{}}
\end{Verbatim}
\end{tcolorbox}

    \textbf{Note}: If you run ariba on your own data then you may need to
edit the above command depending on how your own files are named.

    \hypertarget{ariba-output}{%
\subsection{ARIBA output}\label{ariba-output}}

While you are waiting for ARIBA to run on all 3 samples, go to the
\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-run}{ARIBA
wiki (https://github.com/sanger-pathogens/ariba/wiki/Task:-run} and read
about the ARIBA output and what each of the columns in the report.tsv
output file mean.

    \hypertarget{summarising-the-results}{%
\subsection{Summarising the results}\label{summarising-the-results}}

Now gather together the results

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary all\PYZus{}results ariba.*/report.tsv
\end{Verbatim}
\end{tcolorbox}

    Look at the files produced by ariba summary:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    You should see 3 files:

\texttt{all\_results.csv} \texttt{all\_results.phandango.csv}
\texttt{all\_results.phandango.tre}

Now look at the file all\_results.csv and answer the questions below.

    \hypertarget{exercises}{%
\subsection{Exercises}\label{exercises}}

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Which AMR genes are present in all 3 samples?
\item
  Which AMR genes are absent in sample ERR1067813 but present in the
  other two samples?
\item
  Which AMR genes are absent in sample ERR1067814 but present in the
  other two samples?
\item
  Which AMR genes are absent in sample ERR1067815 but present in the
  other two samples?
\end{enumerate}

    Now go to the next part of the tutorial where we
\href{make_custom_db.ipynb}{prepare a custom reference database for
ARIBA}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{prepare-a-custom-reference-database-for-ariba}{%
\section{Prepare a custom reference database for
ARIBA}\label{prepare-a-custom-reference-database-for-ariba}}

For the \textit{N. gonorrhoeae} data from the ARIBA manuscript, we were
also interested in a looking for a specific set of sequences and SNPs
not found in any of the standard databases. This means using custom
reference data as opposed to one of the public reference sets like in
the previous section.

We were interested in several reference sequences (coding and non-coding
sequences) and particular variants in those sequences. The idea is to
generate input files to ARIBA that contain all the sequences and
variants of interest, using the ARIBA function
\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-aln2meta}{aln2meta}.

First, let's change to the directory with the reference data.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ../Ref/
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{using-one-reference-sequence}{%
\subsection{Using one reference
sequence}\label{using-one-reference-sequence}}

We start by describing \textit{folP}, but the method is (nearly) the same
for all other sequences.

There are several alleles of \textit{folP} and we want to include them all
in the ARIBA database. We have a SNP of interest R228S in the sequence
in the allele called ``folP.WHO\_F\_01537c'', which confers resistance
to sulphonamides. When we run ARIBA, a particular sample may have a
different allele from folP.WHO\_F\_01537c, but we would still like to
know whether it has the SNP R228S. However, the position may not be 228
because of insertion or deletions. So we use a multiple alignment of all
the reference alleles, and just supply the SNP to ARIBA in one of the
alleles, in this case folP.WHO\_F\_01537c.

The ``aln2meta'' function of ARIBA needs two input files: a multiple
alignment file of the alleles, and a tab-delimited file of the SNPs of
interest. In this case, the SNPs file simply contains one line:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat aln2meta\PYZus{}input/folP\PYZus{}in.tsv
\end{Verbatim}
\end{tcolorbox}

    There are four columns:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\tightlist
\item
  Sequence name
\item
  SNP (an amino acid change at position 228, where R is the wild type
  and S is the variant)
\item
  A ``group name'' for this SNP. This is optional and a dot ``.'' means
  no group name. Putting SNPs into the same group allows ARIBA to report
  them together later on.
\item
  A description of the SNP. This will appear in ARIBA's output files to
  save looking up the reason the SNP is of interest.
\end{enumerate}

This file is used together with the mulitple alignment file to generate
input files to ARIBA when making the database. This is the command to
run:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba aln2meta \PYZhy{}\PYZhy{}variant\PYZus{}only aln2meta\PYZus{}input/folP.aln \PY{l+s+se}{\PYZbs{}}
aln2meta\PYZus{}input/folP\PYZus{}in.tsv coding aln2meta\PYZus{}output/folP
\end{Verbatim}
\end{tcolorbox}

    A few things to note about the above command:

\begin{enumerate}
\def\labelenumi{\arabic{enumi}.}
\item
  The option \texttt{-\/-variant\_only} was used, which affects how
  ARIBA reports later on when summarizing across all samples. We are
  only interested in this gene being present if it has a variant that
  causes resistance.
\item
  \texttt{aln2meta\_input/folP.aln} is the name of the multiple
  alignment file.
\item
  The sequence is ``coding'', which makes ARIBA treat it as such, in
  particular it will interpret the variant R228S as an amino acid change
  at position 228 in the translated amino acid sequence. The input
  sequence is still in nucleotides, not amino acids.
\item
  The command outputs three files, which can be used as input to the
  command \texttt{ariba\ prepareref} (see later).
\end{enumerate}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls aln2meta\PYZus{}output/folP*
\end{Verbatim}
\end{tcolorbox}

    Although we have many more reference sequences of interest to deal with
for the complete analysis, for illustrative purposes here we can use the
three files \texttt{aln2meta\_output/folP*} to make an ARIBA reference
database:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba prepareref \PYZhy{}f aln2meta\PYZus{}output/folP.fa \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}m aln2meta\PYZus{}output/folP.tsv \PYZhy{}\PYZhy{}cdhit\PYZus{}clusters \PY{l+s+se}{\PYZbs{}}
    aln2meta\PYZus{}output/folP.cluster test.aribadb
\end{Verbatim}
\end{tcolorbox}

    This made an ARIBA database of just those sequences and the SNP R228S in
a new directory called \texttt{test.aribadb}. Let's check that it was
made:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls test.aribadb
\end{Verbatim}
\end{tcolorbox}

    You do not need to worry about the contents of the new directory
\texttt{test.aribadb}, just know that it can be used as input to run
ARIBA. However, this was for just one of the many reference sequences of
interest, so we will delete it.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
rm \PYZhy{}r test.aribadb
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{using-all-reference-sequences}{%
\subsection{Using all reference
sequences}\label{using-all-reference-sequences}}

We need to deal with each of the reference sequences in turn by running
\texttt{ariba\ aln2meta} on each, like in the above example with
\textit{folP}. The only difference is that some of them are non-coding
sequences, which means that the command must have \texttt{noncoding}
instead of \texttt{coding}. For example:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba aln2meta \PYZhy{}\PYZhy{}variant\PYZus{}only aln2meta\PYZus{}input/16S.aln \PY{l+s+se}{\PYZbs{}}
    aln2meta\PYZus{}input/16S\PYZus{}in.tsv noncoding aln2meta\PYZus{}output/16S
\end{Verbatim}
\end{tcolorbox}

    There are 10 coding sequences and two non-coding sequences. Instead of
writing 12 commands, we will use two `for loops'. First, the coding
sequences:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{k}{for} x in folP gyrA mtrR parC parE penA ponA porB1b rpoB rpsJ
\PY{k}{do}
    ariba aln2meta \PYZhy{}\PYZhy{}variant\PYZus{}only aln2meta\PYZus{}input/\PY{n+nv}{\PYZdl{}x}.aln \PY{l+s+se}{\PYZbs{}}
    aln2meta\PYZus{}input/\PY{n+nv}{\PYZdl{}x}\PY{l+s+se}{\PYZbs{}\PYZus{}}in.tsv coding aln2meta\PYZus{}output/\PY{n+nv}{\PYZdl{}x}
\PY{k}{done}
\end{Verbatim}
\end{tcolorbox}

    And now the two non-coding sequences:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{k}{for} x in 16S 23S
\PY{k}{do}
    ariba aln2meta \PYZhy{}\PYZhy{}variant\PYZus{}only aln2meta\PYZus{}input/\PY{n+nv}{\PYZdl{}x}.aln \PY{l+s+se}{\PYZbs{}}
    aln2meta\PYZus{}input/\PY{n+nv}{\PYZdl{}x}\PY{l+s+se}{\PYZbs{}\PYZus{}}in.tsv noncoding aln2meta\PYZus{}output/\PY{n+nv}{\PYZdl{}x}
\PY{k}{done}
\end{Verbatim}
\end{tcolorbox}

    This has generated three files for each sequence. We will combine these
to make input files for running \texttt{ariba\ prepareref}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat aln2meta\PYZus{}output/*.fa presence\PYZus{}absence/*.fa \PYZgt{} Ngo\PYZus{}ARIBA.fa
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat aln2meta\PYZus{}output/*.tsv presence\PYZus{}absence/presence\PYZus{}absence.tsv \PY{l+s+se}{\PYZbs{}}
    \PYZgt{} Ngo\PYZus{}ARIBA.tsv
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
cat aln2meta\PYZus{}output/*.cluster \PY{l+s+se}{\PYZbs{}}
    presence\PYZus{}absence/presence\PYZus{}absence.clusters \PY{l+s+se}{\PYZbs{}}
    \PYZgt{} Ngo\PYZus{}ARIBA.clusters
\end{Verbatim}
\end{tcolorbox}

    Finally, we have the three input files needed to make a single ARIBA
database that has information on all the sequences and SNPs of interest.
In case the directory is already there, we delete it first, then
generate the database:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
rm \PYZhy{}rf Ngo\PYZus{}ARIBAdb
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba prepareref \PYZhy{}f Ngo\PYZus{}ARIBA.fa \PYZhy{}m Ngo\PYZus{}ARIBA.tsv \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}\PYZhy{}cdhit\PYZus{}clusters Ngo\PYZus{}ARIBA.clusters Ngo\PYZus{}ARIBAdb
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls
\end{Verbatim}
\end{tcolorbox}

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
\PY{n+nb}{cd} ..
\end{Verbatim}
\end{tcolorbox}

    We now have a directory \texttt{Ngo\_ARIBAdb} that can be used as the
reference database when running \texttt{ariba\ run} on each sample.

Now move on to the next part of the tutorial where we
\href{run_ariba.ipynb}{run ARIBA using the custom reference data}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{run-ariba-using-a-custom-reference-database}{%
\section{Run ARIBA using a custom reference
database}\label{run-ariba-using-a-custom-reference-database}}

This section illustrates how to run ARIBA on a large number of samples
using a custom reference database. To save time, you will not actually
run any of the commands in this section (each run of ARIBA takes a few
minutes).

\hypertarget{reference-database}{%
\subsection{Reference database}\label{reference-database}}

First you need a reference database. You will already have one if you
followed the instructions in the previous part of this tutorial
(\href{make_custom_db.ipynb}{Prepare a custom reference data for
ARIBA}).

    \hypertarget{how-to-run-on-one-sample}{%
\subsection{How to run on one sample}\label{how-to-run-on-one-sample}}

ARIBA needs the database directory, which we have called
\texttt{Ngo\_ARIBAdb} in the \href{make_custom_db.ipynb}{previous
section of the tutorial}, and two sequencing reads files
\texttt{reads.1.fastq.gz}, \texttt{reads.2.fastq.gz}. The command to run
ARIBA would look like (do not run this here!):

\begin{verbatim}
ariba run Ngo_ARIBAdb reads.1.fastq.gz reads.2.fastq.gz outdir
\end{verbatim}

The above command will make a new directory called \texttt{outdir} that
contains the results of matching the data for your sample
(reads.1.fastq.gz and reads.2.fastq.gz) against the database
Ngo\_ARIBAdb.

    \hypertarget{run-on-all-samples}{%
\subsection{Run on all samples}\label{run-on-all-samples}}

The \textit{N. gonorrhoeae} dataset consists of 1517 samples, and we need
to run ARIBA on each sample, which can be done with a ``for'' loop.
Assuming that the reads files are named like this:

\begin{verbatim}
ERR1067813.1.fq.gz ERR1067813.2.fq.gz
ERR1067814.1.fq.gz ERR1067814.2.fq.gz
ERR1067815.1.fq.gz ERR1067815.2.fq.gz
\end{verbatim}

Then we could run ARIBA on all samples like this (do not run this
command here!):

\begin{verbatim}
for sample in `ls *.1.fq.gz | sed 's/\.1.fq.gz//'`
do
    ariba run Ngo_ARIBAdb $sample.1.fq.gz $sample.2.fq.gz $sample.ariba
done
\end{verbatim}

The output directory of each sample is called \texttt{\$sample.ariba},
for example ERR1067813.ariba is the output directory for sample
ERR1067813.

Note if you are analysing your own data you may need to edit the command
depending on how your own fastq files are named.

    \hypertarget{ariba-output}{%
\subsection{ARIBA output}\label{ariba-output}}

There is no need to run the commands in this section, to save time they
have already been run and the results are found in the directory
\texttt{ARIBA\_reports}. Take a look:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ARIBA\PYZus{}reports
\end{Verbatim}
\end{tcolorbox}

    Now go to the next part of the tutorial where we
\href{phandango.ipynb}{use Phandango to view the results}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{viewing-ariba-results-in-phandango}{%
\section{Viewing ARIBA results in
Phandango}\label{viewing-ariba-results-in-phandango}}

This section describes how to use
\href{http://phandango.net/}{Phandango} to view a summary of ARIBA
results from many samples.

The most important output file from ARIBA is the report called
\texttt{report.tsv}. For this tutorial, we have all 1517 reports in the
directory \texttt{ARIBA\_reports/}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ARIBA\PYZus{}reports \PY{p}{|} wc \PYZhy{}l
\end{Verbatim}
\end{tcolorbox}

    As we have seen before, ARIBA has a functon called
``\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-summary}{summary}''
that can summarise presence/absence of sequences and/or SNPs across
samples. It takes at least two ariba reports as input, and makes a CSV
file that can be opened in your favourite spreadsheet program, and also
makes input files for Phandango. The two Phandango files (a tree and a
CSV file) can be dropped straight into the Phandango page for viewing.

The tree that ARIBA makes is based on the CSV file, which contains
results of presence/absence of sequence and SNPs, and other information
such as percent identity bewteen contigs and reference sequences. This
means that it does not necessarily represent the real phylogeny of the
samples. It is more accurate to provide a tree built from the sequencing
data. For this reason, we will use a pre-computed tree file
\texttt{tree\_for\_phandango.tre}.

\hypertarget{basic-usage-of-ariba-summary}{%
\subsection{Basic usage of ariba
summary}\label{basic-usage-of-ariba-summary}}

First, let's run \texttt{ariba\ summary} using the default settings,
except we will skip making the tree (option \texttt{-\/-no\_tree}):

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}no\PYZus{}tree out ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    We can see that this made two files:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls out.*
\end{Verbatim}
\end{tcolorbox}

    They are the same except for the first line, which has
Phandango-specific information. ARIBA uses the filenames as sample names
in the output:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head \PYZhy{}n \PY{l+m}{2} out.phandango.csv
\end{Verbatim}
\end{tcolorbox}

    The first name is ``ARIBA\_reports/ERR1067709.tsv'', and the rest are
named similarly. This is not ideal, as it will look ugly in Phandango.
Further, the names must exactly match the names in the tree file for
Phandango to work (have a look in the tree file
\texttt{data/tree\_for\_phandango.tre}). You could do a little hacking
here using the Linux command \texttt{sed} on the CSV file. Or
alternatively, we can supply ARIBA with a file of filenames that also
tells ARIBA what to call the samples in the output CSV files it
produces. Instead of ``data/ARIBA\_reports/ERR1067709.tsv'', we would
like to simply use ``ERR1067709'', which is cleaner and matches the tree
file. It also means we can (and will) repeatedly run
\texttt{ariba\ summary} with different options, and get output files
that can be loaded straight into Phandango. This is one way to make the
file with the naming information:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ls ARIBA\PYZus{}reports/* \PY{p}{|} awk \PYZhy{}F/ \PY{l+s+s1}{\PYZsq{}\PYZob{}print \PYZdl{}0,\PYZdl{}NF\PYZcb{}\PYZsq{}} \PY{p}{|} \PY{l+s+se}{\PYZbs{}}
    sed \PY{l+s+s1}{\PYZsq{}s/.tsv\PYZdl{}//\PYZsq{}} \PYZgt{} filenames.fofn
\end{Verbatim}
\end{tcolorbox}

    The file is quite simple. Column 1 is the filename, and column 2 is the
name we would like to use in the output.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head filenames.fofn
\end{Verbatim}
\end{tcolorbox}

    Now we can rerun summary using this input file. Note the use of the new
option \texttt{-\/-fofn}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}no\PYZus{}tree \PYZhy{}\PYZhy{}fofn filenames.fofn \PY{l+s+se}{\PYZbs{}}
    out ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    Check that the renaming worked:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head \PYZhy{}n \PY{l+m}{2} out.phandango.csv
\end{Verbatim}
\end{tcolorbox}

    Now go to \href{http://phandango.net/}{Phandango http://phandango.net}
and drag and drop the files \texttt{out.phandango.csv} and
\texttt{tree\_for\_phandango.tre} into the window. The result should
like this:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.phandango.default.png}
\caption{Default Layout}
\end{figure}

    This a very high-level summary of the data. For each cluster, it is
simply saying whether or not each sample has a `match'. Green means a
match, and pink means not a match. For presence/absence genes, this
means that the gene must simply be there to count as a match. If it is a
``variant only'' gene, then the gene must be there along with one of the
variants that we told ARIBA about earlier when
\href{make_custom_db.ipynb}{generating the custom ARIBA database}.

    \hypertarget{more-information-per-cluster}{%
\subsection{More information per
cluster}\label{more-information-per-cluster}}

In addition to a simple ``yes'' or ``no'' as to whether a sample
``matches'' a given cluster (as explained above), additional columns can
be output for each cluster. See the
\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-summary}{ARIBA
summary wiki page} for a full description of the options.

Adding more columns can result in a very wide plot, so we will just look
at 23S for now, using the option \texttt{-\/-only\_clusters\ 23S}.
Adding the option \texttt{-\/-preset\ all} will show all available
columns for the 23S cluster:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}only\PYZus{}clusters 23S \PYZhy{}\PYZhy{}preset cluster\PYZus{}all \PYZhy{}\PYZhy{}no\PYZus{}tree \PY{l+s+se}{\PYZbs{}}
  \PYZhy{}\PYZhy{}fofn filenames.fofn out ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    As before, drag and drop the files \texttt{out.phandango.csv} and
\texttt{tree\_for\_phandango.tre} into the window. This time the result
should look like this:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.phandango.23S.cluster_all.png}
\caption{Phandango 23S cluster\_all}
\end{figure}

    Now there are seven columns, showing various findings from ARIBA for
23S. These columns are described in the
\href{https://github.com/sanger-pathogens/ariba/wiki/Task:-summary}{ARIBA
summary wiki page}.

You can control which of the seven cluster columns are output using the
option \texttt{-\/-cluster\_cols} instead of \texttt{-\/-preset}. For
example, this will show just the ``match'' and ``pct\_id'' columns:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}only\PYZus{}clusters 23S \PYZhy{}\PYZhy{}cluster\PYZus{}cols match,pct\PYZus{}id \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}\PYZhy{}no\PYZus{}tree \PYZhy{}\PYZhy{}fofn filenames.fofn out \PY{l+s+se}{\PYZbs{}}
    ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    \hypertarget{variants}{%
\subsection{Variants}\label{variants}}

In the previous screenshot, where the option
\texttt{-\/-preset\ cluster\_all}, there are two variant columns:
``known\_var'' and ``novel\_var''. Green means ``yes'' and pink means
``no''. Here, we are considering a variant to be a difference bewteen
the reference sequence and the assembly contig from the reads. The
known\_var column indicates whether each sample has any variant that is
known to ARIBA, which means it was included when the original ARIBA
database was generated. The novel\_var column indicates whether or not a
sample has any variant that is not already known to ARIBA.

We can view the calls for all the known variants by adding the
\texttt{-\/-known\_variants} option:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}only\PYZus{}clusters 23S \PYZhy{}\PYZhy{}known\PYZus{}variants \PYZhy{}\PYZhy{}no\PYZus{}tree \PY{l+s+se}{\PYZbs{}}
  \PYZhy{}\PYZhy{}fofn filenames.fofn out ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    As before, drag and drop the files \texttt{out.phandango.csv} and
\texttt{tree\_for\_phandango.tre} into the window. This time the result
should like this:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.phandango.23S.known_vars.png}
\caption{Phandango 23S known\_variants}
\end{figure}

    Each known SNP and whether or not it is present is shown, but you may
have noticed there are three colours. Green means ``yes'', orange means
``heterozygous'', and pink means ``no''. \textit{N. gonorrhoeae} has four
copies of 23S. A SNP could be present in none, some, or all copies.
Where it is present in 1,2, or 3 copies, it is called ``heterozygous''
by ARIBA. Consider the SNP 2597CT in column 4, and column 5 to the right
``2597CT.\%''. Samples either do not have this SNP, or have heterozygous
calls. The 2597CT.\% shows the percent of reads that have the SNP, when
mapped to the contig. Hover over the coloured blocks in Phandango to see
the percentage.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}only\PYZus{}clusters 23S \PYZhy{}\PYZhy{}novel\PYZus{}variants \PYZhy{}\PYZhy{}no\PYZus{}tree \PY{l+s+se}{\PYZbs{}}
  \PYZhy{}\PYZhy{}fofn filenames.fofn out ARIBA\PYZus{}reports/*.tsv
\end{Verbatim}
\end{tcolorbox}

    Now Phandango shows all variants found in any of the samples (even if
the variant is unique to one sample). This results in a lot of columns!

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.phandango.23S.novel_vars.png}
\caption{Phandango 23S novel\_variants}
\end{figure}

    Now go to the next part of the tutorial where we
\href{micplot.ipynb}{Investigate MIC data in relation to variants in the
samples}.


    % Add a bibliography block to the postdoc



\newpage





    \hypertarget{investigating-mic-data}{%
\section{Investigating MIC data}\label{investigating-mic-data}}

We will use the \textit{N. gonorrhoeae} dataset again. This tutorial
includes pre-computed output of running ARIBA on all the samples, and
the ARIBA database that was made in the
\href{make_custom_db.ipynb}{earlier section}.

ARIBA has a function called ``micplot'' that generates plots showing the
distribution of MICs across samples with different combinations of
genotypes. To use it, a file is required of MIC data for each sample and
at least one drug. It looks like this:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
head mic\PYZus{}data.tsv
\end{Verbatim}
\end{tcolorbox}

    The first column must be named ``Sample'' and have names that exactly
match those in ARIBA summary files used as input to micplot (we will see
this shortly). The remaining columns should contain drug names and MIC
scores, however, note that the first two columns contain other data that
will be ignored by ARIBA. When ARIBA loads the file, it tries to convert
everything in columns 2 onwards to numbers and assign a value of ``NA''
when this is not possible.

To run micplot, we need an MIC file, like the one above, and an ARIBA
summary file (as described in the \href{phandango.ipynb}{previous
section}). This generates a summary of known 23S and mtrR mutations and
includes the ``assembled'' cluster column, so that interrupted mtrR can
be identified:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba summary \PYZhy{}\PYZhy{}row\PYZus{}filter n \PYZhy{}\PYZhy{}cluster\PYZus{}cols assembled,known\PYZus{}var \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}\PYZhy{}only\PYZus{}clusters 23S,mtrR \PYZhy{}\PYZhy{}v\PYZus{}groups \PYZhy{}\PYZhy{}no\PYZus{}tree \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}\PYZhy{}fofn filenames.fofn summary.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Now we can run micplot using the new file
\texttt{summary.AZMknowngroups.csv} and the MIC file
\texttt{mic\_data.tsv}, showing the MIC data for azithromicin compared
with the different combinations of sequences and known mutations in 23S
and mtrR:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted Azithromycin \PY{l+s+se}{\PYZbs{}}
    mic\PYZus{}data.tsv summary.AZMknowngroups.csv \PY{l+s+se}{\PYZbs{}}
    micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    This produced a pdf file \texttt{micplot.AZMknowngroups.pdf} that looks
like this:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.png}
\caption{micplot AZMknowngroups}
\end{figure}

    There are various options that can be changed. We will show some of them
here, but try running \texttt{ariba\ micplot\ -\/-help} to see all the
options.

\hypertarget{horizontal-lines}{%
\subsection{Horizontal lines}\label{horizontal-lines}}

Horizontal lines can be added to indicate import cutoffs for MIC data,
in this case 0.25 and 2, using the option \texttt{-\/-hlines}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
    Azithromycin mic\PYZus{}data.tsv summary.AZMknowngroups.csv \PY{l+s+se}{\PYZbs{}}
    micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the result:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.hlines.png}
\caption{micplot hlines}
\end{figure}

    \hypertarget{plot-styles}{%
\subsection{Plot styles}\label{plot-styles}}

In the plots above, there is one point per sample. It can be hard to see
how many points there are, despite there being jittering applied to the
horizontal position. We can change the style to group the points
together and plot circles of sizes proportional to the number of
samples, using the option \texttt{-\/-point\_size}. This option
determines the size of the points, but when set to zero if groups the
points together.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
    \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} Azithromycin mic\PYZus{}data.tsv \PY{l+s+se}{\PYZbs{}}
    summary.AZMknowngroups.csv micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the result:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.point_zero.png}
\caption{micplot point size zero}
\end{figure}

    You can choose to not show the violin plots or the dots in the upper
plot, using the option \texttt{-\/-plot\_types}. The default is
\texttt{violin,point}, which means show both. To only show the dots:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
 \PYZhy{}\PYZhy{}plot\PYZus{}types point \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} \PY{l+s+se}{\PYZbs{}}
 Azithromycin mic\PYZus{}data.tsv summary.AZMknowngroups.csv \PY{l+s+se}{\PYZbs{}}
 micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the result:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.dots_only.png}
\caption{micplot dots only}
\end{figure}

    \hypertarget{colours}{%
\subsection{Colours}\label{colours}}

There are various colour options - see the
\href{http://matplotlib.org/users/colormaps.html}{matplotlib colourmaps
page} for all of the available colour palettes. The default is
``Accent'', which has 8 colours. ARIBA will cycle through these,
repeating colours if there are more than 8 columns in the plot. The
palette can be changed using the option \texttt{-\/-colourmap}.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
 \PYZhy{}\PYZhy{}colourmap PiYG \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} Azithromycin mic\PYZus{}data.tsv \PY{l+s+se}{\PYZbs{}}
 summary.AZMknowngroups.csv micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the result:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.PiYG.png}
\caption{micplot PiYG}
\end{figure}

    The palette PiYG is continuous, and is almost white in the middle. This
is not ideal. We can skip the range in the middle, specifically 40-60\%,
using the option \texttt{-\/-colour\_skip}:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
 \PYZhy{}\PYZhy{}colourmap PiYG \PYZhy{}\PYZhy{}colour\PYZus{}skip \PY{l+m}{0}.35,0.65 \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} \PY{l+s+se}{\PYZbs{}}
 Azithromycin mic\PYZus{}data.tsv summary.AZMknowngroups.csv \PY{l+s+se}{\PYZbs{}}
 micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the new plot:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.colour_skip.png}
\caption{micplot colour\_skip}
\end{figure}

    The number of colours can be set to less than the number of columns
using the option \texttt{-\/-number\_of\_colours}. This makes ARIBA
cycle the colours. Here is an example using the first three colours from
the ``Dark2'' colour palette:

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
 \PYZhy{}\PYZhy{}colourmap Dark2 \PYZhy{}\PYZhy{}number\PYZus{}of\PYZus{}colours \PY{l+m}{3} \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} \PY{l+s+se}{\PYZbs{}}
 Azithromycin mic\PYZus{}data.tsv summary.AZMknowngroups.csv \PY{l+s+se}{\PYZbs{}}
 micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    And we only have three colours:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.3colours.png}
\caption{micplot 3 colours}
\end{figure}

    Setting the number of colours to one results in a black and white
figure.

    \begin{tcolorbox}[breakable, size=fbox, boxrule=1pt, pad at break*=1mm,colback=cellbackground, colframe=cellborder]
\prompt{In}{incolor}{ }{\boxspacing}
\begin{Verbatim}[commandchars=\\\{\}]
ariba micplot Ref/Ngo\PYZus{}ARIBAdb/ \PYZhy{}\PYZhy{}interrupted \PYZhy{}\PYZhy{}hlines \PY{l+m}{0}.25,2 \PY{l+s+se}{\PYZbs{}}
 \PYZhy{}\PYZhy{}number\PYZus{}of\PYZus{}colours \PY{l+m}{1} \PYZhy{}\PYZhy{}point\PYZus{}size \PY{l+m}{0} Azithromycin \PY{l+s+se}{\PYZbs{}}
 mic\PYZus{}data.tsv summary.AZMknowngroups.csv micplot.AZMknowngroups
\end{Verbatim}
\end{tcolorbox}

    Here is the black and white figure:

    \begin{figure}
\centering
\includegraphics{Screenshots/screenshot.micplot.AZMknowngroups.blackwhite.png}
\caption{micplot black and white}
\end{figure}

    Congratultaions, you have reached the end of the tutorial!


    % Add a bibliography block to the postdoc



\end{document}
